### PR TITLE
add bloom filter druid expression

### PR DESCRIPTION
**Pull request by clintropolis**
Adds a `bloom_filter_test` Druid expression to allow use in `ExpressionVirtualColumn` and `ExpressionDimFilter`, as well as SQL virtual column expression support.

### CODE REVIEW

File: extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/expressions/BloomFilterExprMacro.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    final Expr arg = args.get(0);
+    final Expr filterExpr = args.get(1);
+
+    if (!filterExpr.isLiteral() || filterExpr.getLiteralValue() == null) {
+      throw new IAE("Function[%s] second argument must be a base64 serialized bloom filter", name());
+    }
+
+
+    final String serializedFilter = filterExpr.getLiteralValue().toString();
+    final byte[] decoded = BaseEncoding.base64().decode(serializedFilter);
```
Comment (asdf2014) [inline extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/expressions/BloomFilterExprMacro.java]: Maybe use `java.util.Base64.getDecoder().decode(serializedFilter)` is a better choice..

FYI, I found some [benchmark](http://java-performance.info/base64-encoding-and-decoding-performance/) results using the initial Java 8 release on the Xeon E5-2650@2.8 Ghz

| Name          | Encode, 100 bytes | Decode, 100 bytes | Encode, 1000 bytes | Decode, 1000 bytes | Encode, 200000000 bytes | Decode, 200000000 bytes |
| ------------- | ----------------- | ----------------- | ------------------ | ------------------ | ----------------------- | ----------------------- |
| JavaXmlImpl   | 0.721 sec         | 1.067 sec         | 0.664 sec          | 0.947 sec          | 0.689 sec               | 0.885 sec               |
| Java8Impl     | 0.808 sec         | 1.101 sec         | 0.712 sec          | 0.913 sec          | 0.699 sec               | 0.887 sec               |
| MiGBase64Impl | 0.887 sec         | 2.113 sec         | 0.788 sec          | 1.978 sec          | 0.812 sec               | 1.928 sec               |
| IHarderImpl   | 1.179 sec         | 2.645 sec         | 1.012 sec          | 2.439 sec          | 0.976 sec               | 2.431 sec               |
| ApacheImpl    | 4.113 sec         | 4.733 sec         | 2.308 sec          | 2.667 sec          | 2.205 sec               | 2.552 sec               |
| GuavaImpl     | 3.305 sec         | 4.178 sec         | 3.12 sec           | 3.755 sec          | 3.102 sec               | 3.744 sec               |
| SunImpl       | 11.153 sec        | 8.281 sec         | 3.992 sec          | 4.533 sec          | 3.289 sec               | 4.096 sec               |

File: extensions-core/druid-bloom-filter/src/test/java/org/apache/druid/query/filter/sql/BloomDimFilterSqlTest.java [hunk 2]
```diff
@@ -112,6 +130,74 @@ public void testBloomFilter() throws Exception
...
+  public void testBloomFilterVirtualColumn() throws Exception
+  {
+    BloomKFilter filter = new BloomKFilter(1500);
+    filter.addString("a-foo");
+    filter.addString("-foo");
+    if (!NullHandling.replaceWithDefault()) {
+      filter.addBytes(null, 0, 0);
+    }
+    byte[] bytes = BloomFilterSerializersModule.bloomKFilterToBytes(filter);
+    String base64 = Base64.encodeBase64String(bytes);
```
Comment (asdf2014) [inline extensions-core/druid-bloom-filter/src/test/java/org/apache/druid/query/filter/sql/BloomDimFilterSqlTest.java]: The same, use `java.util.Base64.getEncoder().encodeToString(bytes)` would be better.

File: extensions-core/druid-bloom-filter/src/test/java/org/apache/druid/query/filter/sql/BloomDimFilterSqlTest.java [hunk 3]
```diff
@@ -112,6 +130,74 @@ public void testBloomFilter() throws Exception
...
+    );
+  }
+
+  @Test
+  public void testBloomFilterVirtualColumnNumber() throws Exception
+  {
+    BloomKFilter filter = new BloomKFilter(1500);
+    filter.addDouble(20.2);
+    byte[] bytes = BloomFilterSerializersModule.bloomKFilterToBytes(filter);
+    String base64 = Base64.encodeBase64String(bytes);
```
Comment (asdf2014) [inline extensions-core/druid-bloom-filter/src/test/java/org/apache/druid/query/filter/sql/BloomDimFilterSqlTest.java]: The same, use `java.util.Base64.getEncoder().encodeToString(bytes)` would be better.

**Comment by asdf2014** [comment]
@clintropolis You are welcome. Maybe we should replace all non JDK Base64 in the project and record it to `druid-forbidden-apis`. I will open a new PR later.

**Comment by clintropolis** [comment]
>Maybe we should replace all non JDK Base64 in the project and record it to druid-forbidden-apis. I will open a new PR later.

I only saw 2 other occurrences I think, both in tests, but probably a good idea to mark forbidden given those benchmarks :+1:

**Comment by clintropolis** [comment]
Nevermind, I see a handful of commons base64 usage in places

**Comment by clintropolis** [comment]
@asdf2014 It would also probably be worth considering driving these conversions through `StringUtils` class to make future changes easier, it already has a `utf8Base64` method that converts a string to a base64 string, so just expanding on that to handle to/from bytes/strings might be appropriate.

**Comment by asdf2014** [comment]
> Instances of {@link Decoder} class are safe for use by multiple concurrent threads.

@clintropolis Yep, I agree with you. In addition, I found the above comment from JDK, so that we can also hold the instance of `Decoder` and `Encoder` in the `StringUtils` class without worrying about the multi-thread problem. (Although, Base64.Encoder#RFC4648 is still a single instance :sweat_smile:)

**Comment by asdf2014** [comment]
@clintropolis I raised up a PR #6913, PTAL.

**Pull request merged**